Bug 1163138 - Add a management command to allow importing perf data

wlach opened the pull request:
<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/536)
<!-- Reviewable:end -->

maurodoglio commented on treeherder/client/thclient/perfherder.py:
```diff
@@ -0,0 +1,105 @@
+    NINETY_DAYS = 7776000
+
+
+class SignatureList(object):
```
This looks more like a dict than a list.
I would suggest to make this class inherit from dict. If you do that you can probably remove __init__, __len__ and __getitem__.

wlach replied:
Yup, good idea. :)

maurodoglio commented on treeherder/client/thclient/perfherder.py:
```diff
@@ -0,0 +1,105 @@
+    def __getitem__(self, key):
+        return self.signatures.get(key)
+
+    def filter(self, *args):
```
Can you add a comment explaining what's the shape of the args items?

wlach replied:
Sounds good, actually added comments all over the place

maurodoglio commented on treeherder/client/thclient/perfherder.py:
```diff
@@ -0,0 +1,105 @@
+
+class Series(object):
+
+    def __init__(self, blob):
```
I guess blob is a list? Maybe you can make it more explicit with a comment?

wlach replied:
Good idea, fixed

maurodoglio commented on treeherder/client/thclient/perfherder.py:
```diff
@@ -0,0 +1,105 @@
+        return property_values
+
+
+class Series(object):
```
This class looks like a smart version of a list, maybe you want to inherit from it?

wlach replied:
Good idea, fixed

maurodoglio commented on treeherder/model/management/commands/import_perf_data.py:
```diff
@@ -0,0 +1,73 @@
+from treeherder.client import PerfherderClient
+from treeherder.model.derived.jobs import JobsModel
+
+import concurrent.futures
```
I don't think we have futures in our requirements. Can you please add it?

wlach replied:
Hmm, it seemed to be installed already on my vagrant instance, maybe something else is already pulling it in. Still, we should add it explicitly I suppose, will do.

edmorley replied:
> maybe something else is already pulling it in.

peep passes --no-deps to pip, so all dependencies have to be specified explicitly.

It's not present in my vagrant environment:
```
(venv)vagrant@local:~$ python
Python 2.7.9 (default, Apr 15 2015, 20:04:59)
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import concurrent.futures
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named concurrent.futures
```

wlach replied:
Maybe I just pip installed it myself and forgot. Will add it regardless.

maurodoglio commented on treeherder/model/management/commands/import_perf_data.py:
```diff
@@ -0,0 +1,73 @@
+            placeholders=signature_property_placeholders,
+            executemany=True)
+
+        pc = PerfherderClient()
```
I'm a bit confused here... why are you storing the series signature directly to the db while the performance series is posted with the Perfherder Client?

maurodoglio replied:
forget it, I misread get_performance_series

maurodoglio commented on treeherder/model/management/commands/import_perf_data.py:
```diff
@@ -0,0 +1,73 @@
+                options['num_workers']) as executor:
+            for signature_hash in signatures.get_signature_hashes():
+                signature_props = signatures[signature_hash]
+                executor.submit(_add_series, project, time_interval,
```
You should check outside of the for loop that the result of the futures returned by executor.submit() don't have any exception, otherwise any error will pass silently.

wlach replied:
Fixed in update

maurodoglio commented:
The flake8 check is failing

maurodoglio commented on treeherder/client/thclient/perfherder.py:
```diff
@@ -0,0 +1,149 @@
+    NINETY_DAYS = 7776000
+
+
+class SignatureList(dict):
```
For consistency maybe you can rename it SignatureCollection?

wlach replied:
Good idea! Will fix.

wlach commented:
Hey @maurodoglio, I think this is good to review again.

maurodoglio commented on treeherder/client/thclient/perfherder.py:
```diff
@@ -0,0 +1,165 @@
+        '''
+        endpoint = self.PERFORMANCE_DATA_ENDPOINT.format(time_interval)
+        for signature in signature_list:
+            endpoint += '&signatures=%s' % signature
```
I don't remember if I mentioned it already, but [requests has a `params` argument](http://docs.python-requests.org/en/latest/user/quickstart/#passing-parameters-in-urls) you can use to build the querystring. And it also supports list of parameters :smiley_cat:

maurodoglio commented on treeherder/model/management/commands/import_perf_data.py:
```diff
@@ -0,0 +1,112 @@
+                try:
+                    future.result()
+                except Exception, e:
+                    print "FAIL: %s" % e
```
In the context of a custom django command [you can write to self.stdout (or self.stderr)](https://docs.djangoproject.com/en/1.8/howto/custom-management-commands/#management-commands-output) instead of printing directly to stdout.

maurodoglio commented on treeherder/model/management/commands/import_perf_data.py:
```diff
@@ -0,0 +1,112 @@
+                    executor.shutdown(wait=False)
+                    for future in futures:
+                        future.cancel()
+                    sys.exit(1)
```
Maybe you can raise a CommandError here? In that way it can be caught when the command is executed using call_command

maurodoglio commented:
I promise those are my last comments :bowtie:

wlach commented:
Ok! Fixed things up as you suggested. Also added unit tests for the perf stuff, which actually turned up a bunch of bugs. :P I think we might be finally good to go here.
